Update `FindNetwork` to address network name duplications

yongtang opened the pull request:
**- What I did**
This fix is part of the effort to address https://github.com/docker/docker/issues/30242#issuecomment-273517205 where issue arise when creating services because of the fact that multiple networks may share the same name (within or across local/swarm scopes).

**- How I did it**

The focus of this fix is to allow creation of service when a network in local scope has the same name as the service network.

**- How to verify it**


An integration test has been added.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

This fix is related to #30242.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented on api/server/router/network/backend.go:
```diff
@@ -15,6 +15,6 @@ type Backend interface {
 	ConnectContainerToNetwork(containerName, networkName string, endpointConfig *network.EndpointSettings) error
 	DisconnectContainerFromNetwork(containerName string, networkName string, force bool) error
-	DeleteNetwork(name string) error
+	DeleteNetwork(networkID string) error
```
Change to `networkID` to enforcing the usage of ID instead of Name/ID/Prefix

yongtang commented on api/server/router/network/network_routes.go:
```diff
@@ -208,7 +139,12 @@ func (n *networkRouter) postNetworkConnect(ctx context.Context, w http.ResponseW
 	}
 
-	return n.backend.ConnectContainerToNetwork(connect.Container, vars["id"], connect.EndpointConfig)
+	// Always make sure there is no ambiguity with respect to the network ID/name
```
It is a lot easier to enforcing non-ambiguity here.

yongtang commented on api/server/router/network/network_routes.go:
```diff
@@ -225,22 +161,31 @@ func (n *networkRouter) postNetworkDisconnect(ctx context.Context, w http.Respon
 
-	return n.backend.DisconnectContainerFromNetwork(disconnect.Container, vars["id"], disconnect.Force)
+	// Always make sure there is no ambiguity with respect to the network ID/name
+	nw, err := n.backend.FindNetwork(vars["id"])
```
Non-ambiguity at the api level.

yongtang commented on api/server/router/network/network_routes.go:
```diff
@@ -208,7 +139,12 @@ func (n *networkRouter) postNetworkConnect(ctx context.Context, w http.ResponseW
+	if err != nil {
+		return err
+	}
+	return n.backend.ConnectContainerToNetwork(connect.Container, nw.ID(), connect.EndpointConfig)
```
Pass `nw.ID()` for connect.

yongtang commented on api/server/router/network/network_routes.go:
```diff
@@ -225,22 +161,31 @@ func (n *networkRouter) postNetworkDisconnect(ctx context.Context, w http.Respon
+	if err != nil {
+		return err
+	}
+	return n.backend.DisconnectContainerFromNetwork(disconnect.Container, nw.ID(), disconnect.Force)
```
Pass nw.ID() for disconnect.

yongtang commented on api/server/router/network/network_routes.go:
```diff
@@ -225,22 +161,31 @@ func (n *networkRouter) postNetworkDisconnect(ctx context.Context, w http.Respon
+			return err
+		}
+	} else {
+		if err := n.backend.DeleteNetwork(nw.ID); err != nil {
```
Delete might happen in both local and scoped networks.

yongtang commented on api/server/router/network/network_routes.go:
```diff
@@ -412,3 +357,82 @@ func (n *networkRouter) postNetworksPrune(ctx context.Context, w http.ResponseWr
 }
+
+func (n *networkRouter) findNetwork(term string) (types.NetworkResource, error) {
+	// NOTE: This findNetwork is differnt from FindNetwork from the daemon.
```
This `findNetwork` is different from `daemon.FindNetwork` as it will be cross scopes (both local and swarm).

dnephin replied:
Could you make this comment a docstring (above the function) ? Right now it looks like it's documenting the variables inside the function.

Also what do you think about `findUniqueNetwork()` as a name?

yongtang replied:
Thanks @dnephin. The comment has been updated to doc string.

yongtang commented on daemon/cluster/executor/container/container.go:
```diff
@@ -544,19 +544,6 @@ func (c *containerConfig) serviceConfig() *clustertypes.ServiceConfig {
-// networks returns a list of network names attached to the container. The
-// returned name can be used to lookup the corresponding network create
-// options.
-func (c *containerConfig) networks() []string {
```
The `networks()` is returning the name only, which should not be encouraged (should always use ID if possible).

yongtang commented on daemon/container_operations.go:
```diff
@@ -1053,3 +1050,10 @@ func (daemon *Daemon) DeactivateContainerServiceBinding(containerName string) er
 	return sb.DisableService()
 }
+
+func getNetworkID(name string, endpointSettings *networktypes.EndpointSettings) string {
```
Alway try to use `NetworkID` whenever `NetworkID` is available.

thaJeztah replied:
This name is a bit odd (from just the name, I would expect it to _always_ return an ID), not a big thing because it's not exported, but perhaps we should find a better name for it

yongtang commented on daemon/network.go:
```diff
@@ -249,7 +267,9 @@ func (daemon *Daemon) createNetwork(create types.NetworkCreateRequest, id string
 	if nw != nil {
 		if create.CheckDuplicate {
-			return nil, libnetwork.NetworkNameError(create.Name)
+			if !agent || nw.Info().Dynamic() {
```
In case of swarm networks, we have to allow creation even if a local network with the same name already exist.

yongtang commented:
The Jenkins failure seems real. Will spend time to investigate it.

yongtang commented:
All tests passed now.

vdemeester commented:
/cc @sanimej @mavenugo @aboch @aaronlehmann

dnephin commented on integration-cli/docker_api_network_test.go:
```diff
@@ -60,14 +60,17 @@ func (s *DockerSuite) TestAPINetworkCreateCheckDuplicate(c *check.C) {
+	id2 := createNetwork(c, configNotCheck, true)
+
+	deleteNetwork(c, id1, true)
+	deleteNetwork(c, id2, true)
```
Those these be `defer`, and called immediately after the create for each?

yongtang replied:
Thanks. The PR has been updated.

dnephin reviewed (commented):
There seem to be a bunch of changes to DeleteNetwork.

Could you give more of an explanation of what this changes, and why it fixes the problem?

yongtang commented:
Thanks @dnephin for the review. The goal of this PR is to try to address the issue caused by name duplication in network names.

Previously, the network name is unique so some of the implementation uses will look up the network name (instead of network ID) when needed.

However, later on the flag `checkDuplicate` was introduced and it is possible now to have multiple networks with the same name (the IDs are still unique) if `checkDuplicate` is `false`.

Adding swarm network in, the same network name could appears in either `local` or `swarm` scope. And even in the same scope (like `swarm` scope), the same network name could be referring to different network IDs.

However, in the existing implementation, many places uses `FindNetwork` to lookup a network based on its name (unfortunately).
 
That causes issues:
1. When deleting a network with a name, it may not always delete the same network.
2. In swarm network, deploy a service may fail (see https://github.com/docker/docker/issues/30242#issuecomment-273517205) because the network ID passed at the beginning may be referring to another network ID when service deployment is actually executed on an individual node.

This PR tries to address the issue mentioned by always avoid `FindNetwork` and use network ID for look up if possible (so that the network referred is unique and consistent).

dnephin commented on daemon/network.go:
```diff
@@ -28,20 +28,38 @@ func (daemon *Daemon) NetworkControllerEnabled() bool {
+// 2. Full Name
+// 3. Partial ID
+// as long as there is no ambiguity
+func (daemon *Daemon) FindNetwork(term string) (libnetwork.Network, error) {
```
Maybe `FindUniqueNetwork()` would be a better name for these, to make it clear that if the term resolves to more than one network it will be an error.

yongtang replied:
Thanks. The name has been changed.

dnephin commented on daemon/network.go:
```diff
@@ -28,20 +28,38 @@ func (daemon *Daemon) NetworkControllerEnabled() bool {
+	if len(listByPartialID) > 1 {
+		return nil, fmt.Errorf("network %s is ambiguous (%d matches found based on ID prefix)", term, len(listByPartialID))
+	}
+	return nil, libnetwork.ErrNoSuchNetwork(term)
```
minor nit: this might read better as a switch/case ?

```go
switch {
case len(listByFullName) == 1:
    return ...
case len(listByFullName) > 1:
    return ...
if len(listByPartialID) == 1:
    return ...
if len(listByPartialID) > 1:
    return ...

```

yongtang replied:
Thanks. Done.

dnephin reviewed (commented):
Design LGTM, thanks for the explanation

vdemeester commented:
ping @mavenugo @aboch

vdemeester commented:
Also ping @yongtang. Needs to rebase and to take care of some @dnephin comments 👼

yongtang commented:
Thanks @dnephin @vdemeester for the review and sorry for the late response. The PR has been updated. Please take a look and let me know if there are any issues.

yongtang commented:
The Jenkins failure seems to be for real. Will take a look and try to get it fixed.

yongtang commented:
The Jenkins failure should have been fixed now. Think the WindowsR1 failure is unrelated.

dnephin commented on integration-cli/docker_cli_swarm_test.go:
```diff
@@ -1888,3 +1888,72 @@ func (s *DockerSwarmSuite) TestSwarmServiceLsFilterMode(c *check.C) {
+
+	// Create Service with the same name
+	out, err := d.Cmd("service", "create", "--name", "top", "--network", name, "busybox", "top")
+	c.Assert(err, checker.IsNil, check.Commentf(out))
```
Using this will give you better error messages when the assertion fails:

```go
result := icmd.RunCmd(d.Command(...))
result.Assert(c, icmd.Success)
```

Same with all the calls below. You can also check for stdout in the same call with:

```go
result.Assert(c, icmd.Expected{
    Out: n3.ID,
})
```

And you can get the stdout from `result.Stdout()`, and stderr from `result.Stderr()`

yongtang commented:
Thanks @dnephin for the review. The PR has been updated with `result := icmd.RunCmd(d.Command(...))..`. Please take a look and let me know if there are any other issues.

dnephin reviewed (approved):
LGTM

vdemeester commented:
@yongtang needs a rebase 👼

yongtang commented:
The issue is complicated. Now in command line `docker service create`, the network name will be converted to network ID (vs. done in the daemon previously):
https://github.com/docker/cli/blob/ab3ea637b6f6ad766a61117f3040b82e45a59dc0/cli/command/service/opts.go#L348-L360

However, as `convertNetworks` use `inspect` to do the network name conversion, a duplication error would be returned in case there are multiple network IDs available for the same name.

In certain situations, it might be possible that there are multiple networks with the same name but with different scope, e.g, one `foo` in local and one `foo` in swarm.

For that I think it makes sense to change the way `convertNetworks` is done so that only networks related to swarm scope are searched.

Will open an issue in https://github.com/docker/cli shortly.

yongtang commented:
Created docker/cli#167 which might be needed to fix this PR.

yongtang commented:
Rebased and all tests passed now.

cpuguy83 commented:
ping @sanimej

shin- commented:
Bump! I believe @sanimej is no longer a maintainer - who else would be able to approve this to have it move forward?

yongtang committed:
Update `FindUniqueNetwork` to address network name duplications

This fix is part of the effort to address 30242 where
issue arise because of the fact that multiple networks
may share the same name (within or across local/swarm scopes).

The focus of this fix is to allow creation of service
when a network in local scope has the same name as the
service network.

An integration test has been added.

This fix fixes 30242.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Rebased to address the merge conflict.

yongtang committed:
Update and use FindNetwork on Windows.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
/cc @thaJeztah @stealthybox

vdemeester commented on api/server/router/network/network_routes.go:
```diff
@@ -518,3 +528,79 @@ func (n *networkRouter) postNetworksPrune(ctx context.Context, w http.ResponseWr
 }
+
+// findUniqueNetwork will search network across different scopes (both local and swarm).
+// NOTE: This findUniqueNetwork is differnt from FindUniqueNetwork from the daemon.
```
s/differnt/different/

yongtang replied:
@vdemeester Updated. Thanks! 🎉 🎊 ❤️

vdemeester reviewed (approved):
LGTM 🐯 (with a small typo)

thaJeztah commented:
oh, @yongtang looks like this needs a rebase 😓

cpuguy83 commented on daemon/network.go:
```diff
@@ -29,36 +29,36 @@ func (daemon *Daemon) NetworkControllerEnabled() bool {
+	case len(listByFullName) == 1:
+		return listByFullName[0], nil
+	case len(listByFullName) > 1:
+		return nil, fmt.Errorf("network %s is ambiguous (%d matches found based on name)", term, len(listByFullName))
```
We should be using `fmt.Errorf` for these errors.

yongtang replied:
@cpuguy83 Sorry, that was probably due to the multiple rebase/conflict that was overlooked. Let me take a look and have a fix for it.

cpuguy83 replied:
No worries. I've got it fixed in my errdefs or.

thaJeztah commented on api/server/router/network/backend.go:
```diff
@@ -12,11 +12,11 @@ import (
 // to provide network specific functionality.
 type Backend interface {
-	FindNetwork(idName string) (libnetwork.Network, error)
+	FindUniqueNetwork(idName string) (libnetwork.Network, error)
```
I wonder if we should've kept the old name here (`FindNetwork`); the issue being resolved here is a bugfix, so I don't see a real reason to rename the function (in general, I'd _expect_ a function like this to return a unique network, so not sure a change in naming is really needed)

@yongtang @vdemeester wdyt?
